hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientInvocationServiceImpl.java

```diff
@@ -287,29 +294,29 @@ public boolean isSmartRoutingEnabled() {
         return isSmartRoutingEnabled;
     }
 
-    private class CleanResourcesTask implements Runnable {
+    private class BackupTimeoutTask implements Runnable {
         @Override
         public void run() {
             for (ClientInvocation invocation : invocations.values()) {
-                ClientConnection connection = invocation.getSendConnection();
-                if (connection == null) {
-                    continue;
-                }
+                invocation.detectAndHandleBackupTimeout(operationBackupTimeoutMillis);
+            }
+        }
+    }
 
-                if (!connection.isAlive()) {
-                    notifyException(invocation, connection);
-                    continue;
-                }
+    private class ConnectionRemovedListener implements ConnectionListener<ClientConnection> {
+        @Override
+        public void connectionAdded(ClientConnection connection) {
 
-                if (isBackupAckToClientEnabled) {
-                    invocation.detectAndHandleBackupTimeout(operationBackupTimeoutMillis);
-                }
-            }
         }
 
-        private void notifyException(ClientInvocation invocation, ClientConnection connection) {
-            Exception ex = new TargetDisconnectedException(connection.getCloseReason(), connection.getCloseCause());
-            invocation.notifyException(ex);
+        @Override
+        public void connectionRemoved(ClientConnection connection) {
+            for (ClientInvocation invocation : invocations.values()) {
+                if (invocation.getPermissionToNotifyForDeadConnection(connection)) {
```

Comment:
Shouldn't we first check if the invocations sent connection is actually the connection that is closed before trying the acquire permission to notify? If the sentConnection of the invocation is something else, then CAS will fail and we will spin until the sentConnection for the invocation becomes null

Reply:
The code is simplified, we don't have the while loop anymore.